File: src/day01/main.rs
use itertools::Itertools;
use shared::puzzle_input;

fn mk_parse_table(s: &str) -> Box<[(usize, &str)]> {
    s.split_ascii_whitespace()
        .enumerate()
        .map(|(i, prefix)| (i % 10, prefix))
        .collect_vec()
        .into_boxed_slice()
}

const P1_PREFICES: &str = "0 1 2 3 4 5 6 7 8 9";
const P2_PREFICES: &str = "0 1 2 3 4 5 6 7 8 9 zero one two three four five six seven eight nine";

fn read_p2(line: &str, parse_table: &[(usize, &'static str)]) -> usize {
    let mut v: Vec<usize> = vec![];
    let mut it = 0..line.len();
    while let Some(i) = it.next() {
        for (digit_value, prefix) in parse_table {
            if line[i..].starts_with(prefix) {
                v.push(*digit_value);
                if prefix.len() > 1 {
                    it.nth(prefix.len() - 2); // skip prefix
                }
                break;
            }
        }
    }
    v[0] * 10 + v[v.len() - 1]
}

fn main() {
    let input = puzzle_input!();
    let lines = input.split_ascii_whitespace();
    let table1 = mk_parse_table(P1_PREFICES);
    let table2 = mk_parse_table(P2_PREFICES);
    let read_p1 = |s| read_p2(s, &table1);
    let read_p2 = |s| read_p2(s, &table2);
    let p1: usize = lines.clone().map(read_p1).sum();
    let p2: usize = lines.map(read_p2).sum();
    println!("{p1}\n{p2}");
}

File: src/day01/main.py
DIGIT_WORDS = list(enumerate("zero one two three four five six seven eight nine".split()))

def p1(line: str):
    xs = [int(c) for c in line if c.isdecimal()]
    return xs[0] * 10 + xs[-1]

def p2Digits(line):
    while len(line):
        c = line[0]
        if c.isdecimal():
            yield int(c)
            line = line[1:]
        else:
            for i, prefix in DIGIT_WORDS:
                if line.startswith(prefix):
                    yield i
                    line = line[len(prefix):]
                    break
            else:
                line = line[1:]

def p2(line):
    xs = list(p2Digits(line))
    return xs[0] * 10 + xs[-1]

def main(input: str):
    print(f'{sum(p1(line) for line in input.splitlines())}')
    print(f'{sum(p2(line) for line in input.splitlines())}')